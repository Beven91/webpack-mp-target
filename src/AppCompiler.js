const DefinePlugin = require("webpack/lib/DefinePlugin");
const PlatformAdapter = require('./PlatformAdapter');
const platforms = require('./platforms');

module.exports = class MpAppCompiler {
  constructor(compiler, options) {
    this.compiler = compiler;
    this.options = {
      from: options.from || 'wx',
      target: options.target || 'wx',
    }
    // 原代码平台类型
    this.fromPlatform = platforms[this.options.from];
    // 编译的目标平台
    this.targetPlatform = platforms[this.options.target];
    // 校验参数
    if (this.validate()) {
      // 初始化
      this.initialize();
    }
  }

  // 获取全局变量名转换配置
  variables() {
    const defines = {};
    // 这里将来源的全局变量 同意替换成目标的全局变量 ，例如: wx ---> tt
    defines[this.fromPlatform.global] = this.targetPlatform.global;
    return defines;
  }

  initialize() {
    const compiler = this.compiler;
    compiler.plugin('emit', (compilation, callback) => {
      const adapter = new PlatformAdapter(compilation, this);
      adapter.transform();
      callback();
    });
    compiler.apply(
      // 将全局变量进行重命名
      new DefinePlugin(this.variables())
    )
  }

  validate() {
    const platformsKeys = Object.keys(platforms).join(',')
    if (!this.fromPlatform) {
      throw new Error(`不支持的原代码类型from:${this.options.from},\n仅支持以下类型\n:${platformsKeys}`);
    } else if (!this.targetPlatform) {
      throw new Error(`不支持的目标类型target:${this.options.target},\n仅支持以下类型\n:${platformsKeys}`);
    }
    return (this.fromPlatform !== this.targetPlatform);
  }
}