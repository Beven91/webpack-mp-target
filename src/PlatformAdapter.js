const path = require('path');
const CssAdapter = require('./adapters/CssAdapter');
const TemplateAdapter = require('./adapters/TemplateAdapter');

module.exports = class PlatformAdapter {
  constructor(compilation, callback, context) {
    this.compilation = compilation;
    this.compiler = context.compiler;
    this.from = context.fromPlatform;
    this.target = context.targetPlatform;
    this.adapters = [
      new CssAdapter(this.from, this.target),
      new TemplateAdapter(this.from, this.target)
    ]
    this.transform();
    callback();
  }

  transform() {
    const compilation = this.compilation;
    const allAssets = compilation.assets;
    const assetKeys = Object.keys(allAssets);
    assetKeys
      .forEach((assetKey) => {
        const ext = path.extname(assetKey).toLowerCase();
        const basename = path.extname(assetKey);
        const asset = compilation.assets[assetKey];
        const adapter = this.adapters.find((adapter) => adapter.fromExt === ext);
        if (adapter) {
          const newAssetKey = basename + adapter.target;
          // 处理asset
          adapter.apply(asset);
          // 删除原有的asset
          delete compilation.assets[assetKey];
          // 重新设置asset
          compilation[newAssetKey] = asset;
        }
      })
  }
}

