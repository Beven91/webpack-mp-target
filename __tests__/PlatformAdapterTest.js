/**
 * @module PlatformAdapterTest
 * @name 测试 PlatformAdapterTest模块
 * @description
 */
import AppCompiler from '../src/AppCompiler';
import MockCompiler from 'webpack/lib/Compiler';
import PlatformAdapter from '../src/PlatformAdapter';
import complilation from '../__mocks__/complication';


describe('PlatformAdapter', () => {
  test('apply', () => {
    const callback = jest.fn();

    const mockCompiler = new MockCompiler();
    const compiler = new AppCompiler(mockCompiler, { target: 'bytedance' });

    const adapter = new PlatformAdapter(complilation, callback, compiler);

    // 匹配complilation镜像
    expect(complilation).toMatchSnapshot();

    // 断言:callback 必须校验
    expect(callback).toHaveBeenCalled();
  });
});
