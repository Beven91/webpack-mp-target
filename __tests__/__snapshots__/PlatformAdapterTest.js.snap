// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlatformAdapter apply 1`] = `
Object {
  ".wxml[object Object]": Object {
    "size": [Function],
    "source": [Function],
  },
  ".wxss[object Object]": Object {
    "size": [Function],
    "source": [Function],
  },
  "assets": Object {
    "a/d/hello.ttss": Object {
      "source": [Function],
    },
  },
}
`;
